### ISSUE TITLE
Noticing any instability?

### ISSUE THREAD

**Issue by bbrtj** [Oele/audio-streamgenerator#12]
After last changes to how streaming works (async, taking samples in batches) I am noticing some instability in audio being generated.

Quite often just before the song is about to change I get a couple hundred miliseconds of silence in the old track. It is not 100% though, more like 50 - 60% of the time.

I'm also pretty sure at least once my streamer died during changing songs / mixing.

Are you noticing something similar? Not sure if this is due to module code or my own asynchronous code that runs it.

**Comment by Oele** [comment]
Yes, I did notice some instability issues around the time when changing tracks. I had the impression that these were buffer underruns at the player side caused by the fact that during mixing there is no audio being output a few seconds. This was always the case, but I get the impression it has gotten a bit worse after the latest changes. 

Are you seeing buffer underruns, or do you have the impression that it is actually outputting silent samples? 

I had one case yesterday where I had to restart the streamer because although it said it was still sending audio to icecast, nothing was being received on the player side. Not sure if this was a problem with the streamer or something with icecast.

**Comment by bbrtj** [comment]
I think these are buffer underruns. The track just stops and resumes after a short pause.

My thinking was that ffmpeg is run with a flag that makes it output in real time, so the mixing process has to actually wait a couple of seconds before its read from the new track is fully satisfied - but if it was a problem, it would happen every time, not randomly.

Surely before latest changes the buffer was being advanced much more gradually. In my setup, I use 250 ms batches, ran with 100 ms window in between. So technically I can only be missing at most 350 ms. Can't tell for sure, but I think the silence moments I observe last a bit longer than that.

**Comment by Oele** [comment]
Have you also measured how long the _mix sub runs? There is no audio output at all while it's mixing. I just checked it on my own stream and it takes between 1.6 and 1.8 seconds. 

I'm not sure (yet) what part of that is starting the new ffmpeg process and wat part is the actual mix, but I get the impression that the actual mixing part is actually a large part of it - it seems to spend 1-2 seconds on the "mixing second <nr>..." steps. 

I can think of 2 ways to 'fix' this: 
- add more buffering on the receiving side to absorb this (for example by with the queue-size etc in icecast). Will add some latency though. 
or
- maybe come up with some creative solution in perl, maybe with threading, which would allow us to keep outputting audio at a steady rate while mixing at the same time. 


Btw, regarding the 'real time' flag in ffmpeg - are you using this on the input/source side or for the output/icecast side? I think it only makes sense on the output side - on the input side it will make these kinds of issues worse.

**Comment by bbrtj** [comment]
You're right, real-time flag is on the output, I remembered that wrong. 

In theory, the mixing code could mix only up to `$batch_size` samples at a time, and then the send loop could continue. It would require the mixing state to be moved away from perl variables into something like object state, which would be quite an overhaul. And I'm not sure how much would it help.

Threading (understood as `threads` module) in perl is not a good direction. Forks, maybe, but then we would have to handle IPC. I think it's best if we can avoid breaking out of the interpreter, less complexity.

Anyway, I'll let it wait for some time - already occupied with something else anyway. Maybe in the meantime some better solutions will come to mind.

**Comment by Oele** [comment]
The _mix sub is currently an "all or nothing" thing. It needs to be able to read and manipulate the whole 'mixable' buffer. For example, it finds the last audible sample, and drops everything after that. It finds the last sample that is 'too loud to mix' and starts the mix right after it. It uses the result of the actual mix to figure out what level of volume adjustment is needed, and then applies it to the already mixed buffer. It adjusts the samples that were 'skipped' throughout the process to let the volume adjustment be gradual. Etc. This can all only be done if it has the whole 'mixable' buffer in memory and it has the liberty to manipulate every sample in that buffer until it's done with the whole process. 

I think this makes it impossible or at least very hard to switch to some kind of "only mix X samples at a time" model without completely changing the way mixing works, and thus without having a huge negative impact on the "mix quality". 

The one thing that makes it very hard to solve this issue in the module is that, as the documentation says: "StreamGenerator has no notion of time". It just outputs audio as fast as it can - which is a lot faster than 'real time' most of the time: it's as fast as ffmpeg can decode the source file. Only when it is mixing, it doesn't output anything at all for 1-2 seconds. It's up to ffmpeg, pv, icecast, the player, or whatever else is on the receiving end, to absorb these drops in output with a sufficiently large buffer. 

If we want to change this, we need to add this 'notion of time' to the module. If our code at least *knows* how fast it is supposed to output audio, we could come up with some kind of solution. It could be with threads, or by forking, and I agree with you, I don't really like to go down those routes - but a simpler solution might be to just regularly write to/flush the output buffer between the steps, or even within the loops, of the _mix sub. Of course, this means that we would need some kind of additional in memory 'flushable' buffer that contains a couple of seconds of audio, before the 'mixable' buffer starts. 

I'm not sure though that this is really worth the effort and additional complexity - isn't it a whole lot easier to just make sure that there is a sufficiently large buffer between the stream generator and the player, that can absorb these hickups? Assuming you are using icecast, the easiest fix would probably be to just increase the `queue-size` in `icecast.xml`. 

The downside of having a larger buffer is that the lag between the moment when you see a new track starting in your log or UI, or the moment you trigger a 'skip' event, and the moment you actually hear the new track, increases. But I'm afraid that that the same thing will happen if we try to solve it with some kind of extra buffer in our code.

**Comment by bbrtj** [comment]
Two additional issues:
- CPU usage is gradually rising over time. On my VPS the daemon starts with 5-7% CPU usage, but after some (long) time it increases to 14-16% (streaming, not mixing). This probably did not happen before
- My streamer just stopped streaming, but its process was still running (for 28h CPU time), as well as icecast. Resetting perl process helped, but no idea why the stream did not reach icecast. I have logs with new song names and it looks okay, so it tried to stream but couldn't. Maybe ffmpeg process from perl to icecast died, forgot to check - but then perl would most likely not work at all? (it wouldn't be able to write to a closed handle)

Edit: the second point could actually be dead ffmpeg. We don't check the return value of `print` in `_send_samples` - currently it will silently return 0 if unsuccessful.

**Comment by Oele** [comment]
Hi @bbrtj 

I've been running 3 streams 24/7 for months now and I haven't seen any buffer underruns or instability anymore. 

In my setup the key was to make sure that `get_new_source` returns as fast as possible - I was doing an `ORDER BY random()` to get a random track from a postgresql table, and this took up to a second sometimes, during which no audio was being processed. 

Are you still having issues?

**Comment by bbrtj** [comment]
Yes, I don't think anything changed. Although I haven't listened my streams much last weeks.

Now that you mentioned database, I checked what kind of logic I have in my `get_new_source`. It's one file slurp, **a couple of directory globs** and some perl logic which should finish quick. So it may be that the globs take some extra time sometimes and I hear silence.

**Comment by bbrtj** [comment]
I have updated the module to the latest version and memoized globbing in my code. I'll listen a bit more tomorrow to see if it still has the underruns. It surely does when skipping, but I guess that's a bit different.

**Comment by bbrtj** [comment]
I got rid of all IO in my streamer and observed no improvements. Kept digging.

I started playing with my event loop configuration. I had 0.25 seconds generated every 0.1 second - though the margin was big enough. Now I'm sitting at 0.5 every 0.05, and I observe no more underruns - even when skipping. Sure, the event loop is less responsive now, but at least it runs nicely.

Now the problem is that 0.25/0.1 ratio is advised in the documentation. I will patch it to match the values I have currently, since they seem to do the trick.
